File: src/main/java/ctn/rcf/gas/GameplayTagJava.java
package ctn.rcf.gas;

import cn.solarmoon.spark_core.gas.GameplayTag;
import com.google.common.collect.ImmutableList;
import cpw.mods.util.Lazy;

import javax.annotation.Nonnull;
import java.util.HashSet;
import java.util.List;

public class GameplayTagJava {
  private final Lazy<GameplayTag> ktObject;
  public final String path;

  public GameplayTagJava(String path) {
    this.path = path;
    this.ktObject = Lazy.of(() -> new GameplayTag(this.path));
  }

  public List<String> parts() {
    return ImmutableList.copyOf(this.path.split("\\."));
  }

  public boolean matchs(@Nonnull final GameplayTagJava other) {
    final List<String> parts = this.parts();
    final List<String> otherParts = other.parts();

    if (otherParts.size() > parts.size()) {
      return false;
    }

    return new HashSet<>(otherParts).containsAll(otherParts.subList(0, parts.size()));
  }

  public boolean matchs(@Nonnull final GameplayTag other) {
    return this.asKotlinObject().matches(other);
  }

  public GameplayTag asKotlinObject() {
    return ktObject.get();
  }

  @Override
  public String toString() {
    return this.path;
  }

  @Override
  public int hashCode() {
    return this.path.hashCode();
  }
}

File: src/main/kotlin/architecture/resonator_combat_framework/gas/GameplayTagJava.kt
package architecture.resonator_combat_framework.gas

import com.google.common.collect.ImmutableList

@JvmRecord
data class GameplayTagJava(val path: String) {
	//	private final Lazy<GameplayTag> ktObject;
	//		this.ktObject = Lazy.of(() -> new GameplayTag(this.path));
	fun parts(): MutableList<String> {
		return ImmutableList.copyOf(this.path.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }
			.toTypedArray())
	}

	fun matchs(other: GameplayTagJava): Boolean {
		val parts = this.parts()
		val otherParts = other.parts()

		if (otherParts.size > parts.size) {
			return false
		}

		return HashSet(otherParts).containsAll(otherParts.subList(0, parts.size))
	}

	//	public boolean matchs(@Nonnull final GameplayTag other) {
	//		return this.asKotlinObject().matches(other);
	//	}
	//	public GameplayTag asKotlinObject() {
	//		return ktObject.get();
	//	}
	override fun toString(): String {
		return this.path
	}
}
